### ISSUE TITLE
jax.scipy.spatial.transform.Rotation.from_quat is missing scalar_first flag

### ISSUE THREAD

**Issue by adam-hartshorne** [jax-ml/jax#25491]
### Description

The Rotation package of methods is supposed to mimic the scipy package. However the from_quat() method is missing the scalar_first flag. The as_quat() has it. 

It also isn't clear from the documtation if the quaternion expected for this function is [x,y,z,w] or [w,x,y,z]

### System info (python version, jaxlib version, accelerator, etc.)

N/A

**Comment by jakevdp** [comment]
Hi - thanks for the report! `jax.scipy.spatial.transform.Rotation` is not under active development, and will be removed as soon as there is a viable alternative (refer to https://jax.readthedocs.io/en/latest/jep/18137-numpy-scipy-scope.html#scipy-spatial for a discussion of this).

That said, we will accept PRs making small changes and/or bug fixes if you're interested in adding support for this keyword.

**Comment by amacati** [comment]
Just wanted to expand on this. I'm doing research in robotics and the rotations API is extremely convenient for us to handle rotations in simulations and controllers. Are you certain you want to drop the support for rotations? I would also be interested in adding functionality to it, namely implementing `__pow__` and `random()`.

Contrary to what is written in the current implementation, we do not need to implement `scipy.stats.special_ortho_group`. Directly working on quaternions yields the same distribution (see the [scipy implementation](https://github.com/scipy/scipy/blob/df134eab5a500c2146ed4552c8674a78d8154ee9/scipy/spatial/transform/_rotation.pyx#L3189) and [this discussion](https://github.com/scipy/scipy/issues/12874)).

Having these functions would be great, especially since implementing them should be limited in scope and it might take some time for scipy to support arbitrary frameworks that support the official Array API.

**Comment by amacati** [comment]
@jakevdp any thoughts?

**Comment by amacati** [comment]
Yes, that would be possible. Though I suspect it would mainly consist of functional jax reimplementations of scipy's Rotation. The question is if that would be worth the time, given that scipy tries to make its functions available for any package implementing the Array interface. It could make more sense to contribute to that effort directly instead of creating another package. Do you know by any chance how much progress the scipy team has made in that direction?

Another idea would be to have a package that adds a custom dtype rotation, backs it internally with a quaternion array, and registers dispatchers for multiplication and broadcasting. Not sure this is a good idea though, since most operations that work fine on other arrays do not work for rotations (__add__, __sub__, __truediv__, ...).

**Comment by amacati** [comment]
Looking at the relevant RFC they state the following:

> When a non-NumPy array type sees compiled code in SciPy (which tends to use the NumPy C API), we have a couple of options:
> 
> (1) dispatch back to the other library (PyTorch, CuPy, etc.).
> (2) convert to a NumPy array when possible (i.e., on CPU via the buffer protocol, DLPack, or __array__), use the compiled code in question, then convert back.
> (3) don't support this at all
> I'll note that (1) is the long-term goal; how to implement this is out of scope for this proposal - for more on that, see the Appendix section. For now we choose to do (2) when possible, and (3) otherwise. Switching from that approach to (1) in the future will be backwards-compatible.
> 

And in the appendix:

> For compiled code, other array types (whether CPU, GPU or distributed) are likely not going to work at all; the SciPy code is written for the NumPy C API. It's not impossible that some Cython code will work with other array types if those support the buffer protocol and the Cython code uses memoryviews - but that's uncommon (won't work at all on GPU, and PyTorch doesn't support the buffer protocol on CPU either).
> 
> There has been a lot of discussion on how this should work. The leading candidate is Uarray, which we already use in scipy.fft (as do matching fft APIs in CuPy and PyFFTW) and has other PRs pending in both SciPy and CuPy. However, there is also resistance to that because it involves a lot of complexity - perhaps too much. So significant work is needed to simplify that. Or switch to another mechanism. This is important work that has to be done, but I'd prefer not to mix that with this proposal.
> 
> Whatever the mechanism, it should work transparently such that scipy.xxx.yyy(x) where x is a non-numpy array should dispatch to the library which implements the array type of x.
> 

Since the rotation functions are implemented in cython and won't be callable with jax tracers, I think the only viable solution is to have functions that scipy can dispatch to according to (3). There are a couple of instances where this is implemented (see e.g. https://github.com/scipy/scipy/pull/20772), but this does require a jax implementation that scipy can dispatch to. Moving rotations out of jax would be detrimental to that approach. One could register these implementations from an external library ofc, but then jax would have to delegate further to the external package.

Another way would be to add generic Array API functions in scipy and those if the calling function is not a numpy array. Do you know if scipy is okay with using the generic Array API as a fallback for non-numpy code?

**Comment by jakevdp** [comment]
Ah, that's too bad. Since the rotation operations are expressible in terms of standard array operations and linear algebra, I'd thought the implementation in terms of the array API would be straightforward. But it seems scipy does have a cython source for these, probably for performance reasons rather than the inability to express the operations in terms of NumPy array ops.

**Comment by amacati** [comment]
Yes. If I recall correctly they used to be implemented with standard numpy functions. Do you know if the scipy team is open to the fallback strategy?

**Comment by jakevdp** [comment]
That would be worth asking in the Array API planning threads. I think I recall some discussion of that sort of approach

**Comment by amacati** [comment]
Yes, I found someone who essentially proposed the same approach in the planning thread. I'll take the discussion over to that thread for now.

**Comment by amacati** [comment]
I have spent the last days exploring possible designs for an Array API implementation for SciPy. My current candidate solution would only use the generic Array API. Furthermore, I intend to write everything in jit-compatible terms. 

One thing that I most likely cannot do on the SciPy side though is registering Rotations as a PyTree. To maintain compatibility with what SciPy currently implements, and because I want to avoid jax-specific concepts in SciPy, I cannot turn the Rotation into a NamedTuple that is supported out of the box. Would you be fine with registering Rotation as PyTree inside jax? I currently only require one line to make them compatible:


```python
from jax.tree_util import register_pytree_node
from scipy.spatial.transform import Rotation as R

register_pytree_node(R, lambda v: ((v._quat,), None), lambda _, c: R(c[0]))
```
